consensus/polybft/stake_manager.go

```diff
@@ -0,0 +1,342 @@
+	output, ok := decoded.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("could not convert decoded outputs to map")
+	}
+
+	outputMap, ok := output["0"].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("could not convert decoded outputs to map")
+	}
+
+	blsKey, ok := outputMap["blsKey"].([4]*big.Int)
+	if !ok {
+		return nil, fmt.Errorf("failed to decode blskey")
+	}
+
+	pubKey, err := bls.UnmarshalPublicKeyFromBigInt(blsKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal BLS public key: %w", err)
+	}
+
+	return &ValidatorMetadata{
+		Address:     address,
+		VotingPower: new(big.Int).SetUint64(stake),
+		BlsKey:      pubKey,
+		IsActive:    true,
+	}, nil
+}
+
+// stakeInfo holds info about validator stake
+type stakeInfo struct {
+	address types.Address
+	stake   uint64
+}
+
+// stakeCOunter sorts and returns stake info for all validators
+type stakeCounter struct {
+	sortedByStake       []*stakeInfo
+	stakeMap            map[types.Address]uint64
+	currentValidatorSet AccountSet
+}
+
+// newStakeCounter returns a new instance of stake counter
+func newStakeCounter(currentValidatorSet AccountSet) *stakeCounter {
+	stakeCounter := &stakeCounter{
+		currentValidatorSet: currentValidatorSet,
+		stakeMap:            make(map[types.Address]uint64, 0),
+	}
+
+	for _, v := range currentValidatorSet {
+		stakeCounter.stakeMap[v.Address] += v.VotingPower.Uint64()
+	}
+
+	return stakeCounter
+}
+
+// addStake adds given amount for a validator to stake map
+func (sc *stakeCounter) addStake(address types.Address, stake uint64) {
+	sc.stakeMap[address] += stake
+}
+
+// removeStake removes given amount for a validator from stake map
+func (sc *stakeCounter) removeStake(address types.Address, stake uint64) {
+	sc.stakeMap[address] -= stake
+}
+
+// sortByStake sorts all validators by their stake amount
+func (sc *stakeCounter) sortByStake() {
+	sc.sortedByStake = make([]*stakeInfo, 0)
+	for a, s := range sc.stakeMap {
+		sc.sortedByStake = append(sc.sortedByStake, &stakeInfo{address: a, stake: s})
+	}
+
+	// sort from highest to lowest stake
+	sort.Slice(sc.sortedByStake, func(i, j int) bool {
+		return sc.sortedByStake[i].stake > sc.sortedByStake[j].stake
+	})
+}
+
+// getStake returns total stake amount for given validator
+func (sc *stakeCounter) getStake(address types.Address) uint64 {
+	return sc.stakeMap[address]
+}
+
+// sortedIndex return the index of the given validator in the sorted list of stakes
+func (sc *stakeCounter) sortedIndex(address types.Address) int {
+	for i, s := range sc.sortedByStake {
+		if s.address == address {
+			return i
+		}
+	}
+
+	return -1
+}
+
+// iterateThroughNewValidators returns a slice of stake info of validators that are not
+// in the current validator set
+func (sc *stakeCounter) iterateThroughNewValidators() []*stakeInfo {
+	newValidators := make([]*stakeInfo, 0)
+
+	for _, s := range sc.sortedByStake {
+		if sc.currentValidatorSet.ContainsAddress(s.address) {
```

Comment:
Maybe it would be optimal to use `func (as AccountSet) GetAddressesAsSet() map[types.Address]struct{} {` and then to check in `O(1)` if address is inside current validator set?

Reply:
Agreed. Done: https://github.com/0xPolygon/polygon-edge/pull/1431/commits/a457027a2a14a5d28b6a92b746d4710a995d5e83